### PR TITLE
CI: Set up circleci to deploy sphinx site via gh-pages

### PR DESCRIPTION
**Pull request by rossbar**
Adds a `deploy-docs` job to the circleci workflow to deploy the tutorials site using `gh-pages`. The doc deployment job will only be run on a merge to master.

The missing component is the ssh fingerprint - this requires the ability to add ssh keys to both the numpy/numpy-tutorials repository on GitHub, as well as adding the private key to circleci. I have the appropriate permissions for circleci but not GitHub. Fortunately, the procedure is well-documented, just follow the instructions here: https://circleci.com/docs/2.0/gh-bb-integration/#creating-a-github-deploy-key. I tested this out on my own fork and everything went smoothly.

Note that this will deploy to the standard URL used for `gh-pages` deployments, something like https://numpy.github.io/numpy-tutorials. Changing the default URL location will be addressed in follow-up.

### CODE REVIEW

**Comment by mattip** [comment]
> Note that this will deploy to the standard URL used for gh-pages deployments, something like https://numpy.github.io/numpy-tutorials.

Because we already have a CNAME entry for DNS in numpy/numpy.github.com to point to numpy.org, any other repo in the numpy org that uses github pages will automatically get a URL like numpy.org/repo-name so nothing else is really needed.

**Comment by mattip** [comment]
> The missing component is the ssh fingerprint 

Hmm, if you use github pages it is easier since the github checkout action handles the authorization. Here is [an example](https://github.com/Quansight/numpy-threading-extensions/blob/main/.github/workflows/push_docs.yml) from another project of the complete workflow. Sorry not to mention it earlier, I meant to do this here too. I guess the advantage of using circle-ci is that you already have the build workflow running.

**Comment by rossbar** [comment]
> I guess the advantage of using circle-ci is that you already have the build workflow running.

Yeah the reason I chose circleci instead of github actions is so that we could have the link to the build artifact in the CI summary. I'm sure something like this is similar with github actions, I just don't know how to do it.

I can create a key pair and configure circleci + add the fingerprint since I have permissions for all that - the only remaining piece would be to have someone with the appropriate github permissions add the public key to this repo.

Though if we want to switch from circleci to github actions that's fine with me!

**Comment by melissawm** [comment]
I like @rossbar 's point about having the built artifacts - they have been important especially since the sphinx build of the myst notebooks has some formatting quirks that only appear in the final built site. We can build it locally but it is certainly more convenient to have it on CI.

@mattip, would you merge this PR once the final CI setup (be it ssh keys or github actions) is done? I have no qualifications to actually review it :)

**Comment by rossbar** [comment]
We talked about this in the triage meeting and @charris will help us by uploading the public key (many thanks!)

One final thing to consider - once this PR is merged, the website will start deploying and be visible at numpy.org/numpy-tutorials. Maybe we should ping the web team to make sure that's okay? We should also make sure that `master` is in a state that's ready for publication!

**Comment by melissawm** [comment]
> We talked about this in the triage meeting and @charris will help us by uploading the public key (many thanks!)
> 
> One final thing to consider - once this PR is merged, the website will start deploying and be visible at numpy.org/numpy-tutorials. Maybe we should ping the web team to make sure that's okay? We should also make sure that `master` is in a state that's ready for publication!

Great point - I'll do that! Thanks!

**Comment by rossbar** [comment]
Converting to draft until the final deploy destination/URL redirection is resolved.

**Comment by mattip** [comment]
Can we do this in two steps: accept this PR to push to the gh-pages branch, and after that is merged find a way to redirect the URL to something more pleasant?

**Comment by rossbar** [comment]
If I understood the slack discussion correctly, the consensus was to avoid the `numpy.org/numpy-tutorials` URL, even if it were only temporary. If that's the case then I don't think we can merge this until the the redirection part is solved.

**Comment by mattip** [comment]
@stefanv was that the intention?

**Comment by mattip** [comment]
@rossbar could you check with @stefanv if the intention was to prevent any use of the numpy-tutorial URL?

**Comment by stefanv** [comment]
I don't think it's a problem to have it hosted at that address temporarily.  I would use `robots.txt` to suppress it being indexed there, set up the CloudFlare (or Netlify) redirect as soon as possible, and not share links to the soon-to-be-invalidated URL.

**Comment by rossbar** [comment]
I've added a meta tag via jinja template in 1680c1a to prevent search engines from indexing the site when deployed at `numpy.org/numpy-tutorials`.

FWIW, the following references made the meta tag seem preferable to `robots.txt`. If there are other ideas about how to achieve this, please LMK!
 - https://developers.google.com/search/docs/advanced/robots/intro
 - https://developers.google.com/search/reference/robots_meta_tag

**Comment by rossbar** [comment]
With that, I think this is ready to go. Once merged, the site should start deploying to the `numpy.org/numpy-tutorials` URL.

**Comment by stefanv** [comment]
@rossbar I'm waiting on CloudFlare credentials, then I can help with the redirect step.

**Comment by stefanv** [comment]
I've now played with the CloudFlare redirects; they work fine, but they only give us the ability to rewrite numpy.org/tutorials to numpy.org/numpy-tutorials.  This is not exactly what we want: we'd like the browser path to remain at numpy.org/tutorials.

Given that folks do not want to see more repositories on github.com/numpy, here are some solutions:

1. Deploy via netlify, who supports [rewrite through status 200 redirects](https://docs.netlify.com/routing/redirects/redirect-options/#http-status-codes)
2. Deploy to a new organization on GitHub, say `numpy-artifacts/tutorials`, with the main website at `numpy-artifacts/numpy.github.io`.
3. Add a worker to CloudFlare; this is not entirely trivial to implement, and limits us to 10k requests per day.  It can be done, as far as I can tell, while supporting caching.

I'd also suggest the order above as the order to try the solutions in, but let's see what others think.

**Comment by mattip** [comment]
I would like to suggest just going with `numpy-tutorials`. If someone wants to work on it, that is fine but it should be lowest priority.

**Review by mattip — APPROVED** [review]
Looks good to me.

**Comment by stefanv** [comment]
I don't understand the objection to the simplest solution either (deploying to `github.com/numpy/tutorials`); given that objection, I consider (1) and (2) above to be relatively little work for a more tidy looking URL scheme.  But, this is not my hill, so I leave it for the rest of you to decide.

**Comment by rgommers** [comment]
I agree with Matti, any of these solutions are not worth the extra complications. I'm perfectly fine with `numpy.org/numpy-tutorials`, the extra 6 characters are a minor thing.

Out of the proposed alternatives, moving everything to Netlify is the only one I consider reasonable. A new GitHub org is much worse than a new repo, and a 10k limit is a nonstarter.

**Comment by stefanv** [comment]
@rgommers If you are open to the Netlify move, I can do that.  Is there any time better than another for downtime?

**Comment by rgommers** [comment]
It doesn't matter from a downtime perspective, the variation in visitor numbers isn't huge, and last time we decided to just do it whenever timing was best to deal with fallout. Sat/Sun is a bit lower than weekdays.

**Comment by rossbar** [comment]
It seems that the netlify route was a dead end. With that, have we settled on `numpy.org/numpy-tutorials` being cemented as the "official" URL for the tutorials? We discussed this at the docs meeting today and it would be nice to start publicizing the excellent work of our tutorial authors - settling on the final URL name is the last major blocker. @mattip @rgommers @stefanv

**Comment by stefanv** [comment]
To document the situation with Netlify: it would have been a good technical solution, but they will only give us a free group account if we link their name either [on the main page or on all other pages](https://answers.netlify.com/t/netlify-open-source-accounts/4769/9).

**Comment by rgommers** [comment]
> With that, have we settled on `numpy.org/numpy-tutorials` being cemented as the "official" URL for the tutorials?

I'd say yes.

**Comment by mattip** [comment]
It seems the next step then is to officially link to numpy.org/numpy-tutorials from the front page of numpy.org and from the numpy/numpy documentation

**Comment by melissawm** [comment]
Thank you so much to all of you that worked on this - I will submit a PR to include the links on numpy.org as soon as we do one last pass at the content and make sure it's all properly done.

**Pull request merged**
